src/garage/_dtypes.py

```diff
@@ -167,3 +166,128 @@ def __new__(cls, env_spec, observations, actions, rewards, terminals,
+            :obj:`env_spec.observation_space`.
+            actions (numpy.ndarray): A numpy array of shape
+            :math:`(A^*)` containing the action for the current time step.
+            These must conform to :obj:`env_spec.action_space`.
+        reward (float): A float representing the reward for taking
+            the action given the observation, at the current time step.
+        terminals (bool): The termination signal for the current time step.
+        env_info (dict): A dict arbitrary environment state
+            information.
+        agent_infos (numpy.ndarray): A dict of arbitrary agent
+            state information. For example, this may contain the hidden states
+            from an RNN policy.
+
+
+    Raises:
+        ValueError: If any of the above attributes do not conform to their
+            prescribed types and shapes.
+
+    """
+
+    def __new__(cls, env_spec, observation, action, reward, next_observation,
+                terminal, env_info, agent_info):  # noqa: D102
+
+        # observations
+        if not env_spec.observation_space.contains(observation):
+            raise ValueError(
+                'observations must conform to observation_space {}, but got '
+                'data with shape {} instead.'.format(
+                    env_spec.observation_space, observation))
+
+        if not env_spec.observation_space.contains(next_observation):
+            raise ValueError(
+                'next_observation must conform to observation_space {},'
+                ' but got data with shape {} instead.'.format(
+                    env_spec.observation_space, next_observation))
+
+        # action
+        if not env_spec.action_space.contains(action):
+            raise ValueError(
+                'action must conform to action_space {}, but got data with '
+                'shape {} instead.'.format(env_spec.action_space, action))
+
+        if not isinstance(agent_info, dict):
+            raise ValueError('Agent_info must be type {}, but got type {} '
+                             'instead.'.format(dict, type(agent_info)))
+
+        if not isinstance(env_info, dict):
+            raise ValueError('Env_info must be type {}, but got type {} '
+                             'instead.'.format(dict, type(env_info)))
+
+        # rewards
+        if not isinstance(reward, float):
+            raise ValueError('Rewards must be type {}, but got type {} '
+                             'instead.'.format(float, type(reward)))
+
+        if not isinstance(terminal, bool):
+            raise ValueError(
+                'Terminal must be dtype bool, but got dtype {} instead.'.
+                format(type(terminal)))
+
+        return super().__new__(Sample, env_spec, observation, action, reward,
+                               next_observation, terminal, env_info,
+                               agent_info)
+
+
+class SamplesBatch(collections.namedtuple('SamplesBatch', [
```

Comment:
i don't see what purpose this type serves -- it's basically `Tuple[List]` which means it's not immutable at all (the inner list is still mutable). If you wanted an immutable sequence of samples you would just use a tuple, but why bother when the elements are immutable?

Please remove this for now and just stick with regular collections of samples. If an interface wants to specify that you can't reorder the samples, it can do that using a tuple. If it doesn't care about order, it can use a list (i.e. regular old python).

Reply:
I agree. I added had added this in reference to a comment you made on the previous PR. I will remove this dtype for now.